### ISSUE TITLE
segfault on entire-thread

### ISSUE THREAD

**Issue by gdkrmr** [neomutt/neomutt#3183]
When running `entire-thread` neomutt crashes with a segfault.

## Expected Behaviour

Show the entire thread

## Actual Behaviour

segfault

Happens on the current manjaro package and master 
## Steps to Reproduce

Bind `entire-thread` to a key and run it.

## How often does this happen?

- Always

## When did it start to happen?

Current master

## NeoMutt Version

the version packaged with Manjaro and current master compiled with 
```
./configure --with-notmuch=1
```

The package that comes with manjaro:
```
$ neomutt -v
NeoMutt 20211029
Copyright (C) 1996-2020 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.15.7-1-MANJARO (x86_64)
ncurses: ncurses 6.3.20211021 (compiled with 6.2.20200212)
libidn2: 2.3.2 (compiled with 2.3.2)
GPGME: 1.16.0
GnuTLS: 3.7.2
libnotmuch: 5.4.0
storage: kyotocabinet, gdbm, bdb, lmdb
compression: lz4, zlib, zstd

Configure options: --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --gpgme --sqlite --autocrypt --lua --notmuch --gss --gnutls --sasl --with-ui=ncurses --with-idn2=/usr --disable-idn --idn2 --bdb --lmdb --kyotocabinet --gdbm --lz4 --zlib --zstd

Compilation CFLAGS: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -I/usr/include/lua5.3 -DNCURSES_WIDECHAR -I/include -I/usr/include/

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +autocrypt +fcntl -flock -fmemopen +futimens +getaddrinfo +gnutls +gpgme +gss 
  +hcache -homespool +idn +inotify -locales_hack +lua -mixmaster +nls +notmuch 
  -openssl +pgp +regex +sasl +smime +sqlite +sun_attachment 

MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```
current master compiled with `./configure --with-notmuch=1`

```
$ ./neomutt -v
NeoMutt 20211029-162-1cabde
Copyright (C) 1996-2020 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.15.7-1-MANJARO (x86_64)
ncurses: ncurses 6.3.20211021 (compiled with 6.3.20211021)
libidn: 1.38 (compiled with 1.38)
libnotmuch: 5.5.0

Configure options: --with-notmuch=1

Compilation CFLAGS: -g -O2 -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I1/include -DNCURSES_WIDECHAR -I/usr/include

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  -autocrypt +fcntl -flock -fmemopen +futimens +getaddrinfo -gnutls -gpgme -gss 
  -hcache -homespool +idn +inotify -locales_hack -lua -mixmaster +nls +notmuch 
  -openssl +pgp +regex -sasl +smime -sqlite +sun_attachment 

MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

```

## Extra Info

* Operating System and its version. Manjaro
* Were you using multiple copies of NeoMutt at once? No
* Were you using 'screen' or 'tmux'? No
* Is your email local (maildir) or remote (IMAP)? Local

### GDB traceback of current master

```
(gdb) bt
#0  0x00005555555abe8d in op_main_entire_thread (shared=0x5555558d4240, priv=0x5555565cabb0, op=<optimized out>) at index/functions.c:2928
#1  0x00005555555aafc6 in mutt_index_menu (dlg=<optimized out>, m_init=<optimized out>) at index/dlg_index.c:1341
#2  0x0000555555575af9 in main (argc=1, argv=<optimized out>, envp=<optimized out>) at main.c:1379
```

There seems to be a check in the source code but for some reason it is not enough:

```
    // If notmuch doesn't contain the message, we're left in an empty
    // vfolder. No messages are found, but nm_read_entire_thread assumes
    // a valid message-id and will throw a segfault.
    //
    // To prevent that, stay in the empty vfolder and print an error.
    if (shared->mailbox->msg_count == 0)
    {
```

**Comment by flatcap** [comment]
Thanks for debugging, @gdkrmr.

`op_main_entire_thread()` is trying to change folder and failing, causing `shared->mailbox` to become `NULL`.
The simple fix is:
```diff
-if (shared->mailbox->msg_count == 0)
+if (!shared->mailbox || (shared->mailbox->msg_count == 0))
```

@Austin-Ray 
Please can you take a look and see if this is enough / the right solution.
Thanks.

**Comment by Austin-Ray** [comment]
@flatcap : That'll fix the seg fault, but I think the error message in that block would be misleading as we're unable to query the database at all. Probably better to do:

```C
if (!shared->mailbox) {
    return IR_ERROR;
}
```

and let neomutt display `Failed to create query, aborting` from `change_folder_notmuch()`

**Comment by flatcap** [comment]
OK.  It's notmuch.  Please can you do it ;-)
Thanks

**Comment by Austin-Ray** [comment]
> OK. It's notmuch. Please can you do it ;-)

Sure

**Comment by Ledarium** [comment]
Any progress on this? @Austin-Ray

**Comment by BeeFox-sys** [comment]
Confirmed still an issue!

**Comment by intr-cx** [comment]
This happens when `nm_default_url` is not set. Same behavior happens when you run `<vfolder-from-query>` and try to use Tab-completion on a tag without `nm_default_url`.

**Comment by flatcap** [comment]
@intr-cx Thanks for the poke.
I _still_ don't use notmuch (and there don't seem to _be_ any other developers)

For now, I've pushed the trivial fix, 36040c0c1,  to the [[release]](https://github.com/neomutt/neomutt/tree/release) branch.
If you're happy to build neomutt, please can you give it a try.

**Comment by intr-cx** [comment]
@flatcap thanks for the quick response. 

`entire-thread` no longer crashes, but the segfault also (still) happens when running `vfolder-from-query`, typing `tag:` and hitting Tab for a list of tags.

**Comment by flatcap** [comment]
@intr-cx I've found and (hopefully) fixed a couple of problems with tag-completion.

If you have a second, please can you try the [[`release`]](https://github.com/neomutt/neomutt/tree/release) branch
80ce65be4 notmuch: fix tag completion crash

**Notes:**
- `release`s history has changed, so you'll may want to delete your local branch and re-pull
- ~I reverted the `:set` bugfix -- having another think~

**Comment by flatcap** [comment]
Update: `release`:

80ce65be4 notmuch: fix tag completion crash
9e09e2cb4 config: fix inherited has_been_set
c4fca575d maildir/notmuch: fix memory leak

**Comment by intr-cx** [comment]
@flatcap I've compiled it and all segfaults look to be resolved. `:set` now shows a list of variables, tab completion no longer shows any results (and doesn't segfault), and `entire-thread` shows nothing (no segfault)!

**Comment by flatcap** [comment]
Brilliant!  Thanks for your help :heart:

> tab completion no longer shows any results

er, is that what you _meant_ to say?

I've created a simple notmuch setup, and tagged some emails.
When I `<vfolder-from-query>` then type `tag:` + <kbd>\<Tab\></kbd>, NeoMutt cycles through the existing tags.

```sh
# List all existing tags
notmuch search --output=tags '*'
```

**Comment by intr-cx** [comment]
@flatcap yes, when neomutt is not aware of the location of notmuch's database, it's not going to show me any tags. That sounds like expected behavior.

**Comment by flatcap** [comment]
Ah, of course!

OK, I'm going to claim this issue is now resolved (and run away).
If you find any other bugs, don't hesitate to open an issue (and pester me :-)

Thanks

**Comment by intr-cx** [comment]
Hmm, after testing a bit further, there is one oddity.

Even though `nm_default_url` is set, tab completion doesn't work until I execute `<vfolder-from-query>` once (which runs successfully). Only after I run the query, I'm able to autocomplete tags in subsequent queries.

**Comment by flatcap** [comment]
Hmm... interesting.
I'm not sure where (and when) the tags come from.

As they could be changed externally, we should get an updated list every time they might be needed:
- `<vfolder-from-query>`
- `<modify-tags>`
- etc

Please can you copy'n'paste this into a new issue.
